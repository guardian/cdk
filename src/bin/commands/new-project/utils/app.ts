import { CodeMaker } from "codemaker";
import type { Imports } from "./imports";
import type { Name } from "./utils";

interface AppBuilderProps {
  imports: Imports;
  appName: Name;
  stack: Name;
  stages: string[];
  outputFile: string;
  outputDir: string;
  comment?: string;
  migrated?: boolean;
}

export class AppBuilder {
  config: AppBuilderProps;

  code: CodeMaker;

  constructor(props: AppBuilderProps) {
    this.config = props;

    this.code = new CodeMaker({ indentationLevel: 2 });
    this.code.closeBlockFormatter = (s?: string): string => s ?? "}";
  }

  async constructCdkFile(): Promise<void> {
    const { comment, outputFile, imports, appName, stack, outputDir, stages } = this.config;

    this.code.openFile(outputFile);
    if (comment) {
      this.code.line(comment);
      this.code.line();
    }

    imports.render(this.code);

    this.code.line("const app = new App();");

    stages.forEach((stage) => {
      this.code.line(
        `new ${appName.pascal}(app, "${appName.pascal}-${stage}", { stack: "${stack.kebab}", stage: "${stage}" });`
      );
    });

    this.code.closeFile(outputFile);
    await this.code.save(outputDir);
  }
}

export const constructApp = async (props: AppBuilderProps): Promise<void> => {
  const builder = new AppBuilder(props);
  await builder.constructCdkFile();
};
