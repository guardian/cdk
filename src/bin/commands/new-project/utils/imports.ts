import type { CodeMaker } from "codemaker";
import type { Name } from "./utils";

interface Import {
  types: string[];
  components: string[];
  basic?: boolean;
}

export class Imports {
  private readonly imports: Record<string, Import>;

  private constructor(imports: Record<string, Import>) {
    this.imports = imports;
  }

  public static newStackImports(addCfnIncludeImports: boolean): Imports {
    return new Imports({
      ...(addCfnIncludeImports && {
        path: {
          types: [],
          components: ["join"],
        },
        "aws-cdk-lib/cloudformation-include": {
          types: [],
          components: ["CfnInclude"],
        },
      }),
      "aws-cdk-lib": {
        types: ["App"],
        components: [],
      },
      "@guardian/cdk/lib/constructs/core": {
        types: ["GuStackProps"],
        components: ["GuStack"],
      },
    });
  }

  public static newAppImports({ kebab, pascal }: Name): Imports {
    return new Imports({
      "aws-cdk-lib": {
        types: [],
        components: ["App"],
      },
      "source-map-support/register": {
        basic: true,
        types: [],
        components: [],
      },
      [`../lib/${kebab}`]: {
        types: [],
        components: [pascal],
      },
    });
  }

  public static newTestImports({ kebab, pascal }: Name): Imports {
    return new Imports({
      "aws-cdk-lib": {
        types: [],
        components: ["App"],
      },
      "aws-cdk-lib/assertions": {
        types: [],
        components: ["Template"],
      },
      [`./${kebab}`]: {
        types: [],
        components: [pascal],
      },
    });
  }

  render(code: CodeMaker): void {
    Object.entries(this.imports)
      // Render "basic" imports before any others, still in alphabetical order
      // Render relative imports after absolute imports
      .sort(([aKey, aImports], [bKey, bImports]) => {
        if (aImports.basic && !bImports.basic) {
          return -1;
        } else if (bImports.basic && !aImports.basic) {
          return 1;
        } else if (aKey.startsWith(".") && !bKey.startsWith(".")) {
          return 1;
        } else if (bKey.startsWith(".") && !aKey.startsWith(".")) {
          return -1;
        } else {
          return aKey.localeCompare(bKey);
        }
      })
      .forEach(([lib, imports]) => {
        imports.basic && code.line(`import "${lib}";`);

        imports.types.length && code.line(`import type { ${imports.types.sort().join(", ")} } from "${lib}";`);

        imports.components.length && code.line(`import { ${imports.components.sort().join(", ")} } from "${lib}";`);
      });
    code.line();
  }
}
