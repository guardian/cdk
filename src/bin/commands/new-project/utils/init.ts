import { copyFileSync, existsSync, lstatSync, mkdirSync, readdirSync, writeFileSync } from "fs";
import { join } from "path";
import { LibraryInfo } from "../../../../constants";

export interface InitConfig {
  outputDir: string;
}

// TODO: Add yarn or npm flag
// TODO: Add project name flag
export class ProjectBuilder {
  templateDir = `${__dirname}/../template`;
  config: InitConfig;

  static validateConfig = (config: InitConfig): void => {
    const path = config.outputDir;
    if (existsSync(path) && readdirSync(path).length > 0) {
      throw new Error(`Directory ${path} is not empty`);
    }
  };

  constructor(config: InitConfig) {
    this.config = config;
  }

  buildDirectory(): void {
    ProjectBuilder.validateConfig(this.config);

    if (!existsSync(this.config.outputDir)) {
      console.log(`Creating ${this.config.outputDir}`);
      mkdirSync(this.config.outputDir);
    }

    console.log("Creating package.json");
    createPackageJson(this.config.outputDir);

    console.log("Copying template files");
    // TODO: Replace any params in files with .template extensions
    this.copyFiles(this.templateDir, this.config.outputDir);

    console.log("Success!");
  }

  copyFiles(sourcePath: string, targetPath: string): void {
    for (const file of readdirSync(sourcePath)) {
      const path = join(sourcePath, file);

      if (path.endsWith(".ignore")) {
        continue;
      } else if (lstatSync(path).isDirectory()) {
        const nestedTargetPath = join(targetPath, file);
        if (!existsSync(nestedTargetPath)) {
          mkdirSync(nestedTargetPath);
        }
        this.copyFiles(path, nestedTargetPath);
      } else if (path.endsWith(".template")) {
        copyFileSync(path, join(targetPath, file.replace(".template", "")));
      } else {
        copyFileSync(path, join(targetPath, file));
      }
    }
  }
}

export const buildDirectory = (config: InitConfig): void => {
  const builder = new ProjectBuilder(config);
  builder.buildDirectory();
};

function createPackageJson(outputDirectory: string): void {
  const { NODE_ENV, CI } = process.env;
  const isTest = NODE_ENV?.toUpperCase() === "TEST" || CI?.toUpperCase() === "TRUE";

  const contents = {
    name: "cdk",
    version: "0.0.0",
    private: true,
    bin: {
      cdk: "bin/cdk.js",
    },
    scripts: {
      build: "tsc",
      watch: "tsc -w",
      test: "jest",
      "test:dev": "jest --watch",
      format: 'prettier --write "{lib,bin}/**/*.ts"',
      cdk: "cdk",
      lint: "eslint lib/** bin/** --ext .ts --no-error-on-unmatched-pattern",
      synth: "cdk synth --path-metadata false --version-reporting false",
      diff: "cdk diff --path-metadata false --version-reporting false",
    },
    devDependencies: {
      "@aws-cdk/assert": LibraryInfo.AWS_CDK_VERSION,
      "@guardian/eslint-config-typescript": "^0.7.0",
      "@types/jest": "^27.0.2",
      "@types/node": "16.11.7",
      "aws-cdk": LibraryInfo.AWS_CDK_VERSION,
      eslint: "^7.32.0",
      jest: "^27.3.1",
      prettier: "^2.4.1",
      "ts-jest": "^27.0.7",
      "ts-node": "^10.4.0",
      typescript: "~4.4.3",
    },
    dependencies: {
      "@aws-cdk/cloudformation-include": LibraryInfo.AWS_CDK_VERSION,
      "@aws-cdk/core": LibraryInfo.AWS_CDK_VERSION,

      /*
      Do not add `@guardian/cdk` to the generated `package.json` file when in TEST as we'll `npm link` it instead.
      See https://docs.npmjs.com/cli/v8/commands/npm-link#caveat

      TODO remove this once the `new` command allows opting out of automatic dependency installation
       */
      ...(!isTest && { "@guardian/cdk": LibraryInfo.VERSION }),

      "source-map-support": "^0.5.20",
    },
  };
  writeFileSync(`${outputDirectory}/package.json`, JSON.stringify(contents, null, 2));
}
