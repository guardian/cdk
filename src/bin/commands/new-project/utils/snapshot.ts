import type { PackageManager } from "aws-sdk/clients/ecr";
import { CodeMaker } from "codemaker";
import type { Imports } from "./imports";
import type { Name } from "./utils";

export interface TestBuilderProps {
  imports: Imports;
  appName: Name;
  stackName: Name;
  outputFile: string;
  outputDir: string;
  comment?: string;
  packageManager: PackageManager;
}

export class TestBuilder {
  config: TestBuilderProps;
  imports: Imports;

  code: CodeMaker;

  constructor(props: TestBuilderProps) {
    this.config = props;
    this.imports = props.imports;

    this.code = new CodeMaker({ indentationLevel: 2 });
    this.code.closeBlockFormatter = (s?: string): string => s ?? "}";
  }

  async constructCdkFile(): Promise<void> {
    this.code.openFile(this.config.outputFile);
    if (this.config.comment) {
      this.code.line(this.config.comment);
      this.code.line();
    }

    this.config.imports.render(this.code);

    this.config.packageManager === "deno" ? this.addTestDeno() : this.addTest();

    this.code.closeFile(this.config.outputFile);
    await this.code.save(this.config.outputDir);
  }

  addTest(): void {
    const { appName, stackName } = this.config;

    this.code.openBlock(`describe("The ${appName.pascal} stack", () =>`);
    this.code.openBlock(`it("matches the snapshot", () =>`);

    this.code.line("const app = new App();");
    this.code.line(
      `const stack = new ${appName.pascal}(app, "${appName.pascal}", { stack: "${stackName.kebab}", stage: "TEST" });`,
    );

    this.code.line(`const template = Template.fromStack(stack);`);
    this.code.line(`expect(template.toJSON()).toMatchSnapshot();`);

    this.code.closeBlock("});");
    this.code.closeBlock("});");
  }

  addTestDeno(): void {
    const { appName, stackName } = this.config;

    this.code.openBlock(`Deno.test("The ${appName.pascal} stack matches the screenshot", async (ctx) =>`);

    this.code.line("const app = new App();");
    this.code.line(
      `const stack = new ${appName.pascal}(app, "${appName.pascal}", { stack: "${stackName.kebab}", stage: "TEST" });`,
    );

    this.code.line(`const template = Template.fromStack(stack);`);
    this.code.line(`await assertSnapshot(ctx, template.toJSON());`);

    this.code.closeBlock("});");
  }
}

export const constructTest = async (props: TestBuilderProps): Promise<void> => {
  const builder = new TestBuilder(props);
  await builder.constructCdkFile();
};
