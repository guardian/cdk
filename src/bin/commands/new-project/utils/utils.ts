import camelCase from "lodash.camelcase";
import upperFirst from "lodash.upperfirst";

export const pascalCase = (str: string): string => {
  return upperFirst(camelCase(str));
};

export interface Name {
  kebab: string;
  pascal: string;
}
