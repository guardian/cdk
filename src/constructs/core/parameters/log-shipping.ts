import type { GuStack } from "../stack";
import { GuStringParameter } from "./base";

export class GuLoggingStreamNameParameter extends GuStringParameter {
  private static instance: GuStringParameter | undefined;

  private constructor(scope: GuStack) {
    super(scope, "LoggingStreamName", {
      description: "SSM parameter containing the Name (not ARN) on the kinesis stream",
      default: "/account/services/logging.stream.name",
      fromSSM: true,
    });
  }

  public static getInstance(stack: GuStack): GuLoggingStreamNameParameter {
    // Resources can only live in the same App so return a new `GuSSMRunCommandPolicy` where necessary.
    // See https://github.com/aws/aws-cdk/blob/0ea4b19afd639541e5f1d7c1783032ee480c307e/packages/%40aws-cdk/core/lib/private/refs.ts#L47-L50
    const isSameStack = this.instance?.node.root === stack.node.root;

    if (!this.instance || !isSameStack) {
      this.instance = new GuLoggingStreamNameParameter(stack);
    }

    return this.instance;
  }
}
