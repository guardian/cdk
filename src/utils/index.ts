export * as array from "./array";
export * as exec from "./exec";
export * as git from "./git";
export * as math from "./math";
export * as singleton from "./singleton";
