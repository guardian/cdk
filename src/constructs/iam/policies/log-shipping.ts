import { isSingletonPresentInStack } from "../../../utils/test";
import { GuLoggingStreamNameParameter } from "../../core/parameters/log-shipping";
import { GuAllowPolicy } from "./base-policy";
import type { GuStack } from "../../core";

export class GuLogShippingPolicy extends GuAllowPolicy {
  private static instance: GuLogShippingPolicy | undefined;

  private constructor(scope: GuStack) {
    super(scope, "GuLogShippingPolicy", {
      actions: ["kinesis:Describe*", "kinesis:Put*"],
      resources: [
        `arn:aws:kinesis:${scope.region}:${scope.account}:stream/${
          GuLoggingStreamNameParameter.getInstance(scope).valueAsString
        }`,
      ],
    });
  }

  public static getInstance(stack: GuStack): GuLogShippingPolicy {
    if (!this.instance || !isSingletonPresentInStack(stack, this.instance)) {
      this.instance = new GuLogShippingPolicy(stack);
    }

    return this.instance;
  }
}
