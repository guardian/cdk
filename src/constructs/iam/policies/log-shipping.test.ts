import "@aws-cdk/assert/jest";
import { SynthUtils } from "@aws-cdk/assert";
import { attachPolicyToTestRole, simpleGuStackForTesting } from "../../../utils/test";
import { GuLogShippingPolicy } from "./log-shipping";
import type { SynthedStack } from "../../../utils/test";

describe("The GuLogShippingPolicy singleton class", () => {
  it("creates a policy restricted to a kinesis stream defined in a parameter", () => {
    const stack = simpleGuStackForTesting();

    const logShippingPolicy = GuLogShippingPolicy.getInstance(stack);
    attachPolicyToTestRole(stack, logShippingPolicy);

    const json = SynthUtils.toCloudFormation(stack) as SynthedStack;
    expect(json.Parameters.LoggingStreamName).toEqual({
      Type: "AWS::SSM::Parameter::Value<String>",
      Default: "/account/services/logging.stream.name",
      Description: "SSM parameter containing the Name (not ARN) on the kinesis stream",
    });

    expect(stack).toHaveResource("AWS::IAM::Policy", {
      PolicyName: "GuLogShippingPolicy981BFE5A",
      PolicyDocument: {
        Version: "2012-10-17",
        Statement: [
          {
            Action: ["kinesis:Describe*", "kinesis:Put*"],
            Effect: "Allow",
            Resource: {
              "Fn::Join": [
                "",
                [
                  "arn:aws:kinesis:",
                  {
                    Ref: "AWS::Region",
                  },
                  ":",
                  {
                    Ref: "AWS::AccountId",
                  },
                  ":stream/",
                  {
                    Ref: "LoggingStreamName",
                  },
                ],
              ],
            },
          },
        ],
      },
    });
  });

  it("will only be defined once in a stack, even when attached to multiple roles", () => {
    const stack = simpleGuStackForTesting();

    const logShippingPolicy = GuLogShippingPolicy.getInstance(stack);
    attachPolicyToTestRole(stack, logShippingPolicy, "MyFirstRole");
    attachPolicyToTestRole(stack, logShippingPolicy, "MySecondRole");

    expect(stack).toCountResources("AWS::IAM::Policy", 1);
    expect(stack).toCountResources("AWS::IAM::Role", 2);
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });

  it("works across multiple stacks", () => {
    const stack1 = simpleGuStackForTesting();
    const stack2 = simpleGuStackForTesting();

    const logShippingPolicy1 = GuLogShippingPolicy.getInstance(stack1);
    const logShippingPolicy2 = GuLogShippingPolicy.getInstance(stack2);

    attachPolicyToTestRole(stack1, logShippingPolicy1);
    attachPolicyToTestRole(stack2, logShippingPolicy2);

    expect(stack1).toCountResources("AWS::IAM::Policy", 1);
    expect(stack1).toCountResources("AWS::IAM::Role", 1);

    expect(stack2).toCountResources("AWS::IAM::Policy", 1);
    expect(stack2).toCountResources("AWS::IAM::Role", 1);
  });
});
