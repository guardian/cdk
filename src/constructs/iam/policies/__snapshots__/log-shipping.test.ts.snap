// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuLogShippingPolicy singleton class will only be defined once in a stack, even when attached to multiple roles 1`] = `
Object {
  "Parameters": Object {
    "LoggingStreamName": Object {
      "Default": "/account/services/logging.stream.name",
      "Description": "SSM parameter containing the Name (not ARN) on the kinesis stream",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "GuLogShippingPolicy981BFE5A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:Describe*",
                "kinesis:Put*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:kinesis:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":stream/",
                    Object {
                      "Ref": "LoggingStreamName",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GuLogShippingPolicy981BFE5A",
        "Roles": Array [
          Object {
            "Ref": "MyFirstRoleAF66ED75",
          },
          Object {
            "Ref": "MySecondRoleB9F57405",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyFirstRoleAF66ED75": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MySecondRoleB9F57405": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
